File: mobile/lib/domain/entities/quiz.dart
/// Quiz domain entities — quiz, question, attempt, result.
///
/// Phase 10C: Maps to quiz engine API (from 9B backend).

class Quiz {
  final String id;
  final String title;
  final String? description;
  final String? subject;
  final String difficulty;
  final int? timeLimitMinutes;
  final int maxAttempts;
  final int questionCount;
  final int totalPoints;
  final String status;

  const Quiz({
    required this.id,
    required this.title,
    this.description,
    this.subject,
    required this.difficulty,
    this.timeLimitMinutes,
    required this.maxAttempts,
    required this.questionCount,
    required this.totalPoints,
    required this.status,
  });
}

class Question {
  final String id;
  final String questionType; // MCQ, TRUE_FALSE, FILL_IN, DRAG_DROP, MATCHING
  final String questionText;
  final String? questionMediaPath;
  final Map<String, dynamic>? options;
  final int points;
  final int order;

  const Question({
    required this.id,
    required this.questionType,
    required this.questionText,
    this.questionMediaPath,
    this.options,
    required this.points,
    required this.order,
  });
}

class QuizAttempt {
  final String id;
  final String quizId;
  final int attemptNo;
  final String? startedAt;
  final String? completedAt;
  final double? score;
  final double? maxScore;
  final String status;

  const QuizAttempt({
    required this.id,
    required this.quizId,
    required this.attemptNo,
    this.startedAt,
    this.completedAt,
    this.score,
    this.maxScore,
    required this.status,
  });
}

class QuizResultResponse {
  final String questionId;
  final String questionType;
  final String questionText;
  final dynamic studentAnswer;
  final dynamic correctAnswer;
  final bool? isCorrect;
  final double? pointsEarned;
  final int points;
  final String? explanation;

  const QuizResultResponse({
    required this.questionId,
    required this.questionType,
    required this.questionText,
    this.studentAnswer,
    this.correctAnswer,
    this.isCorrect,
    this.pointsEarned,
    required this.points,
    this.explanation,
  });
}

class AttemptResult {
  final QuizAttempt attempt;
  final List<QuizResultResponse> responses;

  const AttemptResult({
    required this.attempt,
    required this.responses,
  });
}

/// Content library item (extended from existing ContentItem for teacher browse).
class LibraryItem {
  final String id;
  final String? schoolId;
  final String title;
  final String contentType;
  final String? levelBand;
  final String? language;
  final String? subject;
  final String? description;
  final String origin; // platform / school
  final String status;

  const LibraryItem({
    required this.id,
    this.schoolId,
    required this.title,
    required this.contentType,
    this.levelBand,
    this.language,
    this.subject,
    this.description,
    required this.origin,
    required this.status,
  });
}

/// Teacher's content submission for platform review.
class ContentSubmission {
  final String id;
  final String contentItemId;
  final String contentTitle;
  final String status;
  final String? submittedAt;
  final String? reviewNotes;
  final String? promotedContentId;

  const ContentSubmission({
    required this.id,
    required this.contentItemId,
    required this.contentTitle,
    required this.status,
    this.submittedAt,
    this.reviewNotes,
    this.promotedContentId,
  });
}

/// Assigned content for a student's class.
class AssignedContent {
  final String id;
  final String contentItemId;
  final String title;
  final String contentType;
  final String? subject;
  final String? description;
  final String? progress; // not_started, started, completed
  final String? streamUrl;

  const AssignedContent({
    required this.id,
    required this.contentItemId,
    required this.title,
    required this.contentType,
    this.subject,
    this.description,
    this.progress,
    this.streamUrl,
  });
}

/// Quiz result for parent dashboard.
class QuizResultSummary {
  final String quizTitle;
  final int attemptNo;
  final double? score;
  final double? maxScore;
  final String status;
  final String? completedAt;

  const QuizResultSummary({
    required this.quizTitle,
    required this.attemptNo,
    this.score,
    this.maxScore,
    required this.status,
    this.completedAt,
  });
}

File: backend/app/schemas/lms/quiz.py
"""Quiz Engine Pydantic schemas — Phase 9B.

Request/response models for quiz CRUD, attempts, responses, and analytics.
"""

from __future__ import annotations

import uuid
from typing import Any

from pydantic import BaseModel, Field, field_validator, model_validator

from app.models.curriculum import (
    is_official_subject,
    is_subject_valid_for_level,
    matieres_for_level,
)
from app.models.taxonomy import normalize_difficulty


def _ensure_subject_for_level(level_band: str | None, subject: str | None) -> None:
    """Curriculum cross-field rule: an OFFICIAL matière must be taught at its niveau.

    Custom (school-defined) matières bypass this check; out-of-scope levels too.
    """
    if (
        level_band
        and subject
        and is_official_subject(subject)
        and not is_subject_valid_for_level(subject, level_band)
    ):
        allowed = ", ".join(matieres_for_level(level_band)) or "(none defined yet)"
        raise ValueError(
            f"subject '{subject}' is not taught at level '{level_band}'. "
            f"Allowed matières: {allowed}"
        )


# ---------------------------------------------------------------------------
# Quiz CRUD
# ---------------------------------------------------------------------------
class QuizQuestionInput(BaseModel):
    question_type: str = Field(
        ..., pattern="^(MCQ|TRUE_FALSE|FILL_IN|DRAG_DROP|MATCHING)$"
    )
    question_text: str = Field(..., min_length=1)
    question_media_path: str | None = None
    options: Any | None = None
    correct_answer: Any = Field(...)
    points: int = Field(default=1, ge=0)
    order: int = Field(default=0, ge=0)
    explanation: str | None = None


class QuizCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=300)
    description: str | None = None
    subject: str | None = Field(None, max_length=50)
    subject_other: str | None = Field(None, max_length=120)  # name when subject == "other"
    level_band: str | None = Field(None, max_length=50)
    difficulty: str | None = Field(None, max_length=20)
    time_limit_minutes: int | None = Field(None, ge=0)
    max_attempts: int = Field(default=1, ge=1)
    shuffle_questions: bool = False
    questions: list[QuizQuestionInput] = Field(default_factory=list)

    @field_validator("difficulty", mode="before")
    @classmethod
    def _normalize_difficulty(cls, v: str | None) -> str | None:
        return normalize_difficulty(v)

    @model_validator(mode="after")
    def _check_subject_for_level(self) -> "QuizCreateRequest":
        _ensure_subject_for_level(self.level_band, self.subject)
        return self


class QuizUpdateRequest(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=300)
    description: str | None = None
    subject: str | None = Field(None, max_length=50)
    level_band: str | None = Field(None, max_length=50)
    difficulty: str | None = Field(None, max_length=20)
    time_limit_minutes: int | None = Field(None, ge=0)
    max_attempts: int | None = Field(None, ge=1)

    @field_validator("difficulty", mode="before")
    @classmethod
    def _normalize_difficulty(cls, v: str | None) -> str | None:
        return normalize_difficulty(v)

    @model_validator(mode="after")
    def _check_subject_for_level(self) -> "QuizUpdateRequest":
        _ensure_subject_for_level(self.level_band, self.subject)
        return self
    shuffle_questions: bool | None = None
    questions: list[QuizQuestionInput] | None = None


class QuizQuestionResponse(BaseModel):
    id: str
    question_type: str
    question_text: str
    question_media_path: str | None = None
    options: Any | None = None
    points: int
    order: int
    explanation: str | None = None
    # correct_answer omitted for students, included for creators


class QuizResponse(BaseModel):
    id: str
    school_id: str | None = None
    created_by: str
    title: str
    description: str | None = None
    subject: str | None = None
    level_band: str | None = None
    difficulty: str | None = None
    time_limit_minutes: int | None = None
    max_attempts: int
    shuffle_questions: bool
    status: str
    language: str | None = None
    total_points: int = 0
    question_count: int = 0


class QuizDetailResponse(QuizResponse):
    questions: list[QuizQuestionResponse] = []


# ---------------------------------------------------------------------------
# Attempts + Responses
# ---------------------------------------------------------------------------
class QuizRespondRequest(BaseModel):
    question_id: uuid.UUID
    student_answer: Any


class AttemptResponse(BaseModel):
    id: str
    quiz_id: str
    student_id: str
    attempt_no: int
    started_at: str
    completed_at: str | None = None
    score: float | None = None
    max_score: int
    status: str


class ResponseResultItem(BaseModel):
    question_id: str
    question_type: str
    question_text: str
    student_answer: Any | None = None
    correct_answer: Any | None = None
    is_correct: bool | None = None
    points_earned: float | None = None
    points: int
    explanation: str | None = None


class AttemptResultResponse(BaseModel):
    attempt: AttemptResponse
    responses: list[ResponseResultItem] = []


# ---------------------------------------------------------------------------
# Analytics
# ---------------------------------------------------------------------------
class QuizAnalyticsResponse(BaseModel):
    quiz_id: str
    title: str
    total_attempts: int
    completed_attempts: int
    average_score: float | None = None
    max_score_achieved: float | None = None
    min_score_achieved: float | None = None
    average_percentage: float | None = None
    question_stats: list[dict] = []
